cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,468 @@
+            nn.ReLU(),
+            nn.Flatten(),
+            layer_init(nn.Linear(64 * 7 * 7, 512)),
+            nn.ReLU(),
+        )
+        self.actor = layer_init(nn.Linear(512, envs.single_action_space.n), std=0.01)
+        self.critic = layer_init(nn.Linear(512, 1), std=1)
+
+    def forward(self, x):
+        """Run policy and value functions."""
+        hidden = self.network(x)
+        logits = self.actor(hidden)
+        value = self.critic(hidden).squeeze(-1)
+        return logits, value
+
+    @staticmethod
+    def sample_action(logits):
+        probs = Categorical(logits=logits)
+        action = probs.sample()
+        return action, probs.log_prob(action)
+
+    @staticmethod
+    def action_log_prob(logits, action):
+        probs = Categorical(logits=logits)
+        return probs.log_prob(action), probs.entropy()
+
+    @staticmethod
+    def action_kl_divergence(old_logits, new_logits):
+        old_probs = Categorical(logits=old_logits)
+        new_probs = Categorical(logits=new_logits)
+        return kl_divergence(old_probs, new_probs).mean()
+
+
+def main():
+    args = parse_args()
+    run_name = f"{args.env_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
+    if args.track:
+        import wandb
+
+        wandb.init(
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            sync_tensorboard=True,
+            config=vars(args),
+            name=run_name,
+            monitor_gym=True,
+            save_code=True,
+        )
+    writer = SummaryWriter(f"runs/{run_name}")
+    writer.add_text(
+        "hyperparameters",
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # Fix multiplication precision on Ampere devices (30xx) prior to PyTorch 1.12
+    # https://pytorch.org/docs/stable/notes/cuda.html#tf32-on-ampere
+    torch.backends.cuda.matmul.allow_tf32 = False
+
+    # TRY NOT TO MODIFY: seeding
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    device = torch.device("cuda" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = envpool.make(
+        args.env_id,
+        env_type="gym",
+        num_envs=args.num_envs,
+        # batch_size=args.epoch_size,  # async
+        num_threads=args.envpool_num_threads,
+        episodic_life=True,
+        seed=args.seed,
+    )
+    envs.is_vector_env = True
+    envs.num_envs = args.num_envs
+    envs.single_action_space = envs.action_space
+    envs.single_observation_space = envs.observation_space
+    envs = RecordEpisodeStatistics(envs)
+    # if args.capture_video:
+    #     envs = gym.wrappers.RecordVideo(envs, f"videos/{run_name}")
+    envs = gym.wrappers.NormalizeObservation(envs)
+    envs = gym.wrappers.TransformObservation(envs, lambda obs: np.clip(obs, -3, 3))
```

Comment:
I am a little hesitant to use `gym.wrappers.NormalizeObservation(envs)` in image-based envs. Dividing the pixels by 255 is a well-known normalization technique. `ppo_procgen.py`, for example, uses only the `gym.wrappers.NormalizeReward`.

https://github.com/vwxyzjn/cleanrl/blob/c41bee932ef998fcbbe301c6d91c2328c1c03ab7/cleanrl/ppo_procgen.py#L188-L197

To this end, my suggestion is not to use wrapper if it does not result in performance benefits.

Reply:
well this is meant to reproduce the DNA paper, where they suggest doing exactly that (normalize images to zero mean unity variance, then clip to -3...3)

should I add a comment to explain why, or no?

Reply:
If the paper specifically normalizes like this way then it's fine.